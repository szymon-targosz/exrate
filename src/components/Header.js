import React from 'react'

const Header = () => (
   <header className="header">
      <div className="container">
         <h1 className="header__title">Exrate</h1>
      </div>
   </header>
)

export default Header;
