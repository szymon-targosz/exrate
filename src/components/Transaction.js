import React from 'react';
import { generateAmountMsg } from '../helpers/transactions';

const Transaction = ({ transaction, rate, removeHandler }) => {
   const { name, amount } = transaction;

   return (
      <div className="transaction">
         <div className="transaction__data">
            <h3 className="heading-3 mb-s">{name}</h3>
            <p className="transaction__value">
               {generateAmountMsg(amount)} / {generateAmountMsg(amount * rate, 'pl')}
            </p>
         </div>
         {removeHandler && (
            <button className="btn btn--link transaction__btn" onClick={removeHandler}>
               Remove
            </button>
         )}
      </div>
   );
};

export default Transaction;
