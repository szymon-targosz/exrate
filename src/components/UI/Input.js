import React from 'react';
import PropTypes from 'prop-types';

const Input = ({ elementType, config, cssClasses, changeHandler, options, label, ...rest }) => {
   let inputElem = null;
   let classes = [ 'input__control' ];

   if (cssClasses) classes = classes.concat(cssClasses);

   switch (elementType) {
      case 'input':
         inputElem = <input className={classes.join(' ')} {...config} onChange={changeHandler} {...rest} />;
         break;

      case 'select':
         inputElem = (
            <select className={classes.join(' ')} {...config} onChange={changeHandler} {...rest}>
               {options.map(option => (
                  <option key={option.value} value={option.value}>
                     {option.text}
                  </option>
               ))}
            </select>
         );
         break;

      default:
         inputElem = <input type="text" className={classes.join(' ')} {...config} onChange={changeHandler} {...rest} />;
         break;
   }

   return (
      <div className="input">
         {inputElem}
         {(label && elementType !== 'select') && <label className="input__label" htmlFor={config.id}>{config.placeholder}</label>}
      </div>
   );
};

Input.propTypes = {
   elementType: PropTypes.string,
   cssClasses: PropTypes.oneOfType([ PropTypes.string, PropTypes.array ]),
   changeHandler: PropTypes.func.isRequired,
   config: PropTypes.shape({
      value: PropTypes.oneOfType([ PropTypes.string, PropTypes.number ]).isRequired
   })
};

export default Input;
