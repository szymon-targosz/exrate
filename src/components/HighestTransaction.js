import React from 'react';
import { connect } from 'react-redux';
import { getHighest, selectTransactions } from '../helpers/transactions';
import Transaction from './Transaction';
import SectionHeader from './UI/SectionHeader';

export const HighestTransaction = ({ transaction, rate }) => {
   const jsx = (
      <section className="highest">
         <SectionHeader title="Highest transaction" />
         <Transaction transaction={transaction} rate={rate} />
      </section>
   );

   return transaction ? jsx : null;
};

const mapStateToProps = ({ transactions, rate, filters }) => ({
   transaction: getHighest(selectTransactions(transactions, filters)),
   rate
});

export default connect(mapStateToProps)(HighestTransaction);
