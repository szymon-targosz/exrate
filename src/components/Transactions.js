import React from 'react';
import { connect } from 'react-redux';
import { removeTransaction, removeAllTransactions } from '../store/actions';
import Transaction from './Transaction';
import TransactionsTotal from './TransactionsTotal';
import { getTotal } from '../helpers/transactions';
import SectionHeader from './UI/SectionHeader';
import { selectTransactions } from '../helpers/transactions';

export const Transactions = ({ removeTransaction, removeAllTransactions, transactions, rate, total }) => (
   <section className="transactions">
      <SectionHeader cssClasses="transactions__header" title="Transactions">
         <button className="btn btn--link" onClick={removeAllTransactions} disabled={!(transactions.length > 0)}>
            Remove All
         </button>
      </SectionHeader>
      <div className="transactions__content">
         <div className="transactions__list">
            {transactions.map(transaction => (
               <Transaction
                  key={transaction.id}
                  removeHandler={() => removeTransaction(transaction.id)}
                  transaction={transaction}
                  rate={rate}
               />
            ))}
            {transactions.length === 0 && <p className="transactions__info p-m">There is no transaction</p>}
         </div>
         <TransactionsTotal total={total} rate={rate} quantity={transactions.length} />
      </div>
   </section>
);

const mapStateToProps = ({ transactions, rate, filters }) => {
   const selected = selectTransactions(transactions, filters);
   const total = getTotal(selected);
   return { transactions: selected, rate, total };
};

const mapDispatchToProps = dispatch => ({
   removeTransaction: id => dispatch(removeTransaction(id)),
   removeAllTransactions: () => dispatch(removeAllTransactions())
});

export default connect(mapStateToProps, mapDispatchToProps)(Transactions);
