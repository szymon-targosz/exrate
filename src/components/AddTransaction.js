import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addTransaction } from '../store/actions';
import SectionHeader from './UI/SectionHeader';
import Input from './UI/Input';

export class AddTransaction extends Component {
   state = { name: '', amount: '', error: undefined };
   inputNameRef = React.createRef();

   handleSubmit = e => {
      e.preventDefault();
      const name = this.state.name.trim();
      if (name === '') return this.setState({ error: 'Please enter the transaction name' });

      const isOccupied = this.props.transactions.find(t => t.name === name);
      if (isOccupied) return this.setState({ error: 'Name already taken' });

      this.props.addTransaction(name, this.state.amount);
      this.setState({ error: undefined, name: '', amount: '' });
      this.inputNameRef.current.focus();
   };

   onNameChange = e => {
      this.setState({ name: e.target.value });
   };

   onAmountChange = e => {
      const amount = e.target.value;
      if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
         this.setState({ amount });
      }
   };

   render() {
      return (
         <section className="add-transaction">
            <SectionHeader title="Add transaction" />
            <form onSubmit={this.handleSubmit} className="p-s">
               <Input
                  elementType="input"
                  changeHandler={this.onNameChange}
                  config={{
                     name: 'name',
                     ref: this.inputNameRef,
                     value: this.state.name,
                     placeholder: 'Name',
                     autoComplete: 'off',
                     id: 'name-input',
                     type: 'text'
                  }}
                  label
               />
               <Input
                  elementType="input"
                  changeHandler={this.onAmountChange}
                  config={{
                     name: 'amount',
                     value: this.state.amount,
                     placeholder: 'Amount (EUR)',
                     autoComplete: 'off',
                     id: 'amount-input',
                     type: 'text'
                  }}
                  label
               />
               {this.state.error && <p className="error">{this.state.error}</p>}
               <button className="btn btn--block" disabled={!this.state.name || !this.state.amount}>
                  Add
               </button>
            </form>
         </section>
      );
   }
}

const mapStateToProps = ({ transactions }) => ({ transactions });

const mapDispatchToProps = dispatch => ({
   addTransaction: (name, amount) => dispatch(addTransaction(name, amount))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddTransaction);
