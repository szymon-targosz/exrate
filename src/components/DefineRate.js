import React, { Component } from 'react';
import { connect } from 'react-redux';
import SectionHeader from './UI/SectionHeader';
import InfoBox from './UI/InfoBox';
import Input from './UI/Input';
import { setExchangeRate } from '../store/actions';
import { generateAmountMsg } from '../helpers/transactions';

export class DefineRate extends Component {
   state = { rate: 0 };

   handleInput = e => {
      const rate = e.target.value;
      if (!rate || rate.match(/^\d{1,}(\.\d{0,9})?$/)) {
         this.setState({ rate: rate });
         this.props.setExchangeRate(rate || 0);
      }
   };

   render() {
      return (
         <section className="define-rate">
            <SectionHeader title="Set exchange rate" cssClasses="define-rate__header" />
            <div className="define-rate__form p-s">
               <Input
                  elementType="input"
                  cssClasses="input__control--short"
                  changeHandler={this.handleInput}
                  config={{
                     value: this.state.rate,
                     placeholder: 'Exchange rate',
                     id: 'rate-input',
                     type: 'text'
                  }}
                  label
               />
            </div>
            <div className="define-rate__current">
               <InfoBox
                  description="Current exchange rate"
                  data={`${generateAmountMsg(1)} = ${generateAmountMsg(parseFloat(this.state.rate) || 0, 'pl', true)}`}
               />
            </div>
         </section>
      );
   }
}

const mapDispatchToProps = dispatch => ({ setExchangeRate: rate => dispatch(setExchangeRate(rate)) });

export default connect(null, mapDispatchToProps)(DefineRate);
