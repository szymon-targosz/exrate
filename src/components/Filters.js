import React, { Component } from 'react';
import SectionHeader from './UI/SectionHeader';
import Input from './UI/Input';
import { connect } from 'react-redux';
import { setTextFilter, sortAlpha, sortAsc, sortDesc } from '../store/actions';

export class Filters extends Component {
   onTextChange = e => {
      this.props.setTextFilter(e.target.value);
   };

   onSortChange = e => {
      const sort = e.target.value;
      if (sort === 'alpha') {
         this.props.sortAlpha();
      } else if (sort === 'desc') {
         this.props.sortDesc();
      } else {
         this.props.sortAsc();
      }
   };

   render() {
      return (
         <section className="filters">
            <SectionHeader title="Filters" />
            <div className="p-s">
               <Input
                  elementType="input"
                  config={{
                     placeholder: 'Search',
                     value: this.props.filters.text,
                     id: 'search-input',
                     type: 'search',
                     autoComplete: 'off'
                  }}
                  changeHandler={this.onTextChange}
                  label
               />
               <Input
                  elementType="select"
                  cssClasses="input__control--select"
                  config={{ value: this.props.filters.sort }}
                  changeHandler={this.onSortChange}
                  options={[
                     { value: 'alpha', text: 'Alphabetically' },
                     { value: 'asc', text: 'Ascending' },
                     { value: 'desc', text: 'Descending' }
                  ]}
               />
            </div>
         </section>
      );
   }
}

const mapStateToProps = ({ filters }) => ({ filters });

const mapDispatchToProps = dispatch => ({
   setTextFilter: text => dispatch(setTextFilter(text)),
   sortAlpha: () => dispatch(sortAlpha()),
   sortAsc: () => dispatch(sortAsc()),
   sortDesc: () => dispatch(sortDesc())
});

export default connect(mapStateToProps, mapDispatchToProps)(Filters);
