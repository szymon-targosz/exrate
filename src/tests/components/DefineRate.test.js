import React from 'react';
import { DefineRate } from '../../components/DefineRate';
import { shallow } from 'enzyme';
import Input from '../../components/UI/Input';

let wrapper, setExchangeRate;

beforeEach(() => {
   setExchangeRate = jest.fn();
   wrapper = shallow(<DefineRate setExchangeRate={setExchangeRate} />);
});

describe('COMPONENT - DefineRate', () => {
   test('Should render DefineRate correctly', () => {
      expect(wrapper).toMatchSnapshot();
   });

   test('Should handle input(rate) change', () => {
      const value = '32.1';
      wrapper.find(Input).prop('changeHandler')({ target: { value } });
      expect(wrapper.state('rate')).toBe(value);
      expect(setExchangeRate).toHaveBeenCalledWith(value);
   });
   
   test('Should set rate to 0 if input is an empty string', () => {
      const value = '';
      wrapper.find(Input).prop('changeHandler')({ target: { value } });
      expect(wrapper.state('rate')).toBe(value);
      expect(setExchangeRate).toHaveBeenCalledWith(0);
   });
});
