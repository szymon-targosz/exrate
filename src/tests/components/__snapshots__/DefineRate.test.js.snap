// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT - DefineRate Should render DefineRate correctly 1`] = `
<section
  className="define-rate"
>
  <SectionHeader
    cssClasses="define-rate__header"
    title="Set exchange rate"
  />
  <div
    className="define-rate__form p-s"
  >
    <Input
      changeHandler={[Function]}
      config={
        Object {
          "id": "rate-input",
          "placeholder": "Exchange rate",
          "type": "text",
          "value": 0,
        }
      }
      cssClasses="input__control--short"
      elementType="input"
      label={true}
    />
  </div>
  <div
    className="define-rate__current"
  >
    <InfoBox
      data="1,00 â‚¬ = 0,00 PLN"
      description="Current exchange rate"
    />
  </div>
</section>
`;
